File: frontend/src/services/translationService.js
/**
 * Translation Service
 * Handles text translation with caching and fallback support
 * Uses backend API for translation, with local fallback support
 */

const API = "http://127.0.0.1:8000";
const TRANSLATION_CACHE = {};

/**
 * Translate text to target language
 * @param {string} text - Text to translate
 * @param {string} targetLanguage - Target language code (en, te, hi, ml)
 * @param {string} sourceLanguage - Source language (default: en)
 * @returns {Promise<string>} - Translated text
 */
export const translateText = async (text, targetLanguage, sourceLanguage = "en") => {
  // Skip if target language is same as source
  if (targetLanguage === sourceLanguage) {
    return text;
  }

  // Check cache first
  const cacheKey = `${text}_${sourceLanguage}_${targetLanguage}`;
  if (TRANSLATION_CACHE[cacheKey]) {
    return TRANSLATION_CACHE[cacheKey];
  }

  try {
    // Call backend translation API
    const response = await fetch(`${API}/translate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        text,
        source_language: sourceLanguage,
        target_language: targetLanguage,
      }),
    });

    if (!response.ok) {
      console.warn(`Translation failed for "${text}"`);
      return text; // Fallback to original text
    }

    const data = await response.json();
    const translatedText = data.translated_text || text;

    // Cache the result
    TRANSLATION_CACHE[cacheKey] = translatedText;
    return translatedText;
  } catch (error) {
    console.error("Translation API error:", error);
    return text; // Fallback to original text
  }
};

/**
 * Batch translate multiple texts
 * @param {Array<string>} texts - Array of texts to translate
 * @param {string} targetLanguage - Target language code
 * @returns {Promise<Array<string>>} - Array of translated texts
 */
export const translateBatch = async (texts, targetLanguage) => {
  if (targetLanguage === "en") {
    return texts; // Skip translation for English
  }

  try {
    const response = await fetch(`${API}/translate-batch`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        texts,
        target_language: targetLanguage,
      }),
    });

    if (!response.ok) {
      return texts; // Fallback
    }

    const data = await response.json();
    return data.translated_texts || texts;
  } catch (error) {
    console.error("Batch translation error:", error);
    return texts;
  }
};

/**
 * Clear translation cache (useful for testing or manual refresh)
 */
export const clearTranslationCache = () => {
  Object.keys(TRANSLATION_CACHE).forEach((key) => delete TRANSLATION_CACHE[key]);
};

/**
 * Get translation status
 */
export const getTranslationCacheSize = () => {
  return Object.keys(TRANSLATION_CACHE).length;
};

File: backend/app/services/translation_service.py
"""
Translation Service for FastAPI
Handles text translation using googletrans library
Provides caching to avoid repeated API calls
"""

from functools import lru_cache
import logging
from typing import List, Dict
from googletrans import Translator, LANGUAGES

logger = logging.getLogger(__name__)

# Initialize translator
translator = Translator()

# Language code mapping
LANGUAGE_MAP = {
    "en": "en",
    "te": "te",  # Telugu
    "hi": "hi",  # Hindi
    "ml": "ml",  # Malayalam
}

# Cache for translations to avoid repeated API calls
@lru_cache(maxsize=1000)
def translate_text_cached(text: str, target_lang: str, source_lang: str = "en") -> str:
    """
    Translate text with caching
    
    Args:
        text: Text to translate
        target_lang: Target language code
        source_lang: Source language code (default: English)
    
    Returns:
        Translated text
    """
    if not text or target_lang == source_lang:
        return text
    
    if target_lang not in LANGUAGE_MAP:
        logger.warning(f"Unsupported language: {target_lang}")
        return text
    
    try:
        result = translator.translate(text, src_language=LANGUAGE_MAP[source_lang], dest_language=LANGUAGE_MAP[target_lang])
        return result['text'] or text
    except Exception as e:
        logger.error(f"Translation error: {e}")
        return text


def translate_batch(texts: List[str], target_lang: str, source_lang: str = "en") -> List[str]:
    """
    Translate multiple texts at once
    
    Args:
        texts: List of texts to translate
        target_lang: Target language code
        source_lang: Source language code (default: English)
    
    Returns:
        List of translated texts
    """
    if target_lang == source_lang or target_lang == "en":
        return texts
    
    return [translate_text_cached(text, target_lang, source_lang) for text in texts]


def get_supported_languages() -> Dict[str, str]:
    """Get dictionary of supported languages"""
    return {code: name for code, name in LANGUAGES.items() if code in LANGUAGE_MAP.values()}


def clear_translation_cache():
    """Clear the translation cache (useful for testing or manual refresh)"""
    translate_text_cached.cache_clear()
